### ISSUE TITLE
Add source "ReaLearn REAPER action"

### ISSUE THREAD

**Issue by helgoboss** [helgoboss/helgobox#649]
Then we could trigger ReaLearn stuff by pressing a toolbar button. I guess there should be a constant number of freely assignable ReaLearn actions. Their meaning depends on the currently open ReaLearn instances.

Use case: Triggering things like saving mapping snapshots.

**Comment by helgoboss** [comment]
I think these constant actions should be named neutrally at all times (it would probably be also very difficult to name them dynamically). The first 8 or 16 actions we add shouldn't support CC control and also not on/off. As such, they should be trigger-only control elements and be named something like "Trigger ...".

We could also add togglable actions and CC-controllable ones later, which would be slightly (but not much) more effort.

In future, it could be interesting to let mappings expose actions dynamically under user-defined command names. Then it would be possible to name those actions in very specific ways and make them available only in certain projects. But that would be much more effort.

**Comment by helgoboss** [comment]
Possibly better idea: Compartment Actions

- Allow each instance to define Actions (each action gets a name and an ID)
- These actions are saved with the compartment and are therefore even saved in compartment presets.
- In addition, we have a new "ReaLearn Action" source (analogously to "ReaLearn Parameter").

This design has the benefit that actions can be named, be saved as part of presets, can have proper feedback (report on/off) and are very local. Each project can define its own set of actions. Disabling ReaLearn instances should unregister their actions.

**Comment by helgoboss** [comment]
Invoking mappings (that's why it will be a source) ... and hence everything that you write (except switching on/off tracing because there's no such target ATM).

**Comment by vonglan** [comment]
So you have a REAPER action "Trigger ReaLearn Action" which can trigger arbitrary ReaLearn mappings, using their ID as parameter?

**Comment by helgoboss** [comment]
No, there's no way to pass such "IDs" when executing REAPER actions.

Each user-defined compartment action will automatically become visible in REAPER's action list (as long as that ReaLearn instance is active). Then you can simply execute it. Main use case: Creating toolbar buttons for it.

**Comment by helgoboss** [comment]
They will disappear but their key bindings will stay (otherwise it would be a pity). In this regard, they will behave exactly like existing ReaLearn actions (search for "ReaLearn" in the action list). The difference is that they will be user-definable.

With this feature, it will essentially be possible to create project-specific toolbars.

**Comment by cldswm** [comment]
> Possibly better idea: Compartment Actions
> 
> * Allow each instance to define Actions (each action gets a name and an ID)
> * These actions are saved with the compartment and are therefore even saved in compartment presets.
> * In addition, we have a new "ReaLearn Action" source (analogously to "ReaLearn Parameter").
> 
> This design has the benefit that actions can be named, be saved as part of presets, can have proper feedback (report on/off) and are very local. Each project can define its own set of actions. Disabling ReaLearn instances should unregister their actions.

This would be a great idea, I just made a FR and also had similar ideaI think this would work great !
